### PR TITLE
Implementation of the STAMP model; Add SBR datasets description to examples/README

### PR DESCRIPTION
**Pull request by bokang-ugent**
The descriptions are added to examples/matching/README.md. Ranking metrics obtained by the NARM model is also added.

### CODE REVIEW

**Comment by bokang-ugent** [comment]
This is an implementation of the STAMP (Short-Term Attention/Memory Priority Model for Session-based Recommendation, [Liu et. al., KDD 2018](https://dl.acm.org/doi/10.1145/3219819.3219950)) model. The mode is also one of the earlier proposed neural model in the line of SBR research, often used as a baseline. 

Changes
* __STAMP__ model is introduced in a new file `models/matching/stamp.py`. A new line that imports the model is added in `models/maching/__init__.py`. The implementation follows directly from the paper, with a few features refer to the original [TensorFlow implementation](https://github.com/uestcnlp/STAMP) as they are not fully clear from the paper (e.g., the normalization of attention weights is not described in the paper).
* __Example script__ `examples/matching/run_sbr.py` is updated to allow STAMP model to be evaluated on the benchmark datasets.
* __README__ results are also updated to `examples/matching/README.md`.

Results

|               | Recall@20 | MRR@20 |
|---------------|-----------|--------|
| YOOCHOOSE1/64 | 0.6675    | 0.2859 |
| YOOCHOOSE1/4  | 0.7079    | 0.3074 |
| DIGINETICA    | 0.5578    | 0.2303 |

* The results are similar to the scores reported in the original paper, with DIGINETICA slightly off. This could be: a. the preprocessing of DIGINETICA in this implementation deviates from the existing implementations (notice the NARM is also a bit worse than the results reported original paper on DIGINETICA ); b. the architecture reported in the paper is not exactly the one that is used in the paper, see [issues](https://github.com/uestcnlp/STAMP/issues) reported the TensorFlow implementation. 

Future work
* Tune Hyperparameter on a validation set. This could potentially improve the quantitative metrics.
* Closely follow the TensorFlow implementation could also potentially improve the scores.
* Consult multiple preprocessing pipelines from different SBR paper, and implement the processing logic that agreed by the majority.

**Review by morningsky — APPROVED** [review]
